text_based_browser.py

```diff
@@ -0,0 +1,98 @@
+import requests
+import os
+import sys
+from collections import deque
+from bs4 import BeautifulSoup
+import colorama
+
+colorama.init()
+
+pages_stack = deque()
+downloaded_pages = {}
+
+
+def save_file(path, name):
+    if name not in downloaded_pages:
```

Comment:
What does this check guard from? Every time we reach here we have already downloaded the page, no?